MLton/MPL doesn't like the `strlen` and `strcpy` imports for error handling

The generated code uses `_import "strlen" public : ...` and `_import "strcpy" public : ...` to manipulate strings returned by `futhark_context_get_error`.

However, the MLton/MPL FFI isn't happy with these imports:
```
/tmp/filehOTM72.0.c:199:13: error: conflicting types for ‘strcpy’
 PUBLIC void strcpy (Objptr x1, CPointer x0);
             ^~~~~~
In file included from /home/ec2-user/proj/mpl/hybrid-sched/build/lib/mlton/include/c-chunk.h:17:0,
                 from /tmp/filehOTM72.0.c:2:
/usr/include/string.h:121:14: note: previous declaration of ‘strcpy’ was here
 extern char *strcpy (char *__restrict __dest, const char *__restrict __src)
              ^~~~~~
/tmp/filehOTM72.0.c:200:14: error: conflicting types for ‘strlen’
 PUBLIC Int64 strlen (CPointer x0);
              ^~~~~~
In file included from /home/ec2-user/proj/mpl/hybrid-sched/build/lib/mlton/include/c-chunk.h:17:0,
                 from /tmp/filehOTM72.0.c:2:
/usr/include/string.h:384:15: note: previous declaration of ‘strlen’ was here
 extern size_t strlen (const char *__s)
               ^~~~~~
MLton [mpl] 20230509.184946-g6f2c989b8 raised: Fail: Process.waits: child 10592 failed with exit status 1
make: *** [main.mpl.bin] Error 1
```

An immediate workaround would be to also generate a small `.c` file that wraps these functions with new names, but that's a bit awkward. Not sure what the best solution is yet.

--- a/src/smlfut.sml
+++ b/src/smlfut.sml
@@ -133,18 +133,40 @@ fun mkShape (info: array_info) v =
 Our solution is to allocate an SML string, copy the C string into it,
 then free the C string.
  *)
-val error_check = fundef "error_check" ["(err,ctx)"]
+
+val strlen =
+  ["fun strlen (p: MLton.Pointer.t) ="]
+  @
+  map indent
+  [ "let"
+  , indent "fun loop i ="
+  , indent (indent ("if MLton.Pointer.getWord8 (p, i) = 0w0 then i else loop (i+1)"))
+  , "in"
+  , indent "loop 0"
+  , "end"
+  ]
+
+val strcpy =
+  [ "fun strcpy (p: MLton.Pointer.t) : string ="
+  , indent "CharVector.tabulate (strlen p, fn i =>"
+  , indent "Char.chr (Word8.toInt (MLton.Pointer.getWord8 (p, i))))"
+  ]
+
+val error_check =
+  ["local"] @ map indent strlen @ map indent strcpy @ ["in"]
+  @
+  fundef "error_check" ["(err,ctx)"]
   [ "if err = 0 then () else"
   , "let val p = "
     ^ fficall "futhark_context_get_error" [("ctx", "futhark_context")] pointer
-  , "val n = " ^ fficall "strlen" [("p", pointer)] "Word64.word"
-  , "val s = " ^ apply "CharVector.tabulate" ["Word64.toInt n", "fn _ => #\" \""]
+  , "val s = strcpy p"
   , "in"
-  , fficall "strcpy" [("s", "string"), ("p", pointer)] pointer ^ ";"
   , fficall "free" [("p", pointer)] "unit" ^ ";"
   , "raise error s"
   , "end"
   ]
+  @
+  ["end"]
 
 fun generateEntryDef manifest (name, ep as entry_point {cfun, inputs, outputs}) =
   let

